File: python/patterns/prototype.py
import copy
from dataclasses import dataclass
from typing import Dict, List


@dataclass
class Character:
    name: str
    skills: List[str]
    stats: Dict[str, int]

    def clone(self) -> "Character":
        return copy.deepcopy(self)


def run() -> None:
    print("\nPrototype")
    original = Character("Hero", ["jump", "run"], {"hp": 100})
    clone = original.clone()
    clone.name = "Hero 2"
    clone.skills.append("fly")

    print(f"Original: {original}")
    print(f"Clone: {clone}")

File: csharp/Patterns/Prototype.cs
using System;
using System.Collections.Generic;

namespace DesignPatterns.Patterns;

public static class Prototype
{
    private class Character
    {
        public Character(string name, List<string> skills, Dictionary<string, int> stats)
        {
            Name = name;
            Skills = skills;
            Stats = stats;
        }

        public string Name { get; set; }
        public List<string> Skills { get; }
        public Dictionary<string, int> Stats { get; }

        public Character Clone()
        {
            return new Character(Name, new List<string>(Skills), new Dictionary<string, int>(Stats));
        }

        public override string ToString()
        {
            return $"{Name} with skills [{string.Join(", ", Skills)}] and hp {Stats["hp"]}";
        }
    }

    public static void Run()
    {
        Console.WriteLine("\nPrototype");
        var original = new Character("Hero", new List<string> { "jump", "run" }, new Dictionary<string, int> { ["hp"] = 100 });
        var clone = original.Clone();
        clone.Name = "Hero 2";
        clone.Skills.Add("fly");

        Console.WriteLine($"Original: {original}");
        Console.WriteLine($"Clone: {clone}");
    }
}
